### ISSUE TITLE
[CT-3232] [Bug] Generic Tests with Refs Run for Their Refs

### ISSUE THREAD

**Issue by jaysobel** [dbt-labs/dbt-core#8867]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Applications of generic tests that contain a reference to a 'supporting' model (in practice often 'helpers' like `dim_numbers` or `dim_date`) are triggered by `dbt test -s supporting_model`.

```
tests/generic/no_date_gaps.sql          with ref('dim_date')
models/dim_date.sql
models/dim_daily_metrics.sql
models/dim_daily_metrics.yml            with test `no_date_gaps` on `dim_daily_metrics`

$ dbt test -s dim_date
  1 of 1 START test no_date_gaps_dim_daily_metrics [RUN]
  1 of 1 ERROR test no_date_gaps_dim_daily_metrics [ERROR]
    Database Error in test no_date_gaps_dim_daily_metrics
    Object 'PROD.CORE.dim_daily_metrics' does not exist or not authorized.
```

This is weird behavior because "generic test" are meant to be applied to other models. While the test implies a dependency between `dim_date -> dim_daily_metrics` it also creates the reverse dependency `dim_daily_metrics -> dim_date` since the instance of the generic test `no_gaps_dim_daily_metrics` is executed.

I haven't proven it, but this pattern could result in a clean slate `dbt build` erroring because the test step for `dim_date` (likely early in the DAG) includes the test of `dim_daily_metrics` (likely later in the DAG).

### Expected Behavior

Custom tests are intended to contain references, and be counted among the tests of those references, and positioned in the DAG accordingly.

Instances of generic tests that contain references should *not* be counted among the tests of those contained references.

Put another way, if a generic test `no_gaps.sql` contains a `ref('dim_date')` and is applied to a model `dim_daily_metrics` then `dbt test -s dim_date` should not trigger the test `no_gaps_dim_daily_metrics`.

### Steps To Reproduce

1. Define a generic test in the "[special subfolder](https://docs.getdbt.com/guides/best-practices/writing-custom-generic-tests#generic-tests-with-standard-arguments)" `tests/generic/generic_test.sql` with a reference to a helper model `ref('helper_model')`
2. Add the generic test to some other model `dim_metrics.sql`.
3. Run `dbt test -s helper_model`
4. Observe that the test `generic_test_dim_metrics` is executed.

### Environment

```markdown
- OS: macOS 13.5.2 (22G91)
- Python: 3.9
- dbt: 1.63
```

### Which database adapter are you using with dbt?

Snowflake

**Comment by dbeatty10** [comment]
Thanks for reaching out about this @jaysobel

If I'm reading your report correctly, what you described actually sounds like the behavior that we'd expect!

### Key insight

Thank for you for sharing this piece:

> the test implies a dependency between `dim_date -> dim_daily_metrics` it also creates the reverse dependency `dim_daily_metrics -> dim_date` since the instance of the generic test `no_gaps_dim_daily_metrics` is executed.

They key insight is that there is no dependency created between the two models. Rather, the test is dependent on each of the two models.

### How to do what you want

When executing `dbt test`, there is a setting called [indirect selection](https://docs.getdbt.com/reference/node-selection/test-selection-examples?indirect-selection-mode=empty#indirect-selection) that determines which test nodes will be executed.

1. `eager` (default) - include ANY test that references the selected nodes
2. `cautious` - restrict to tests that ONLY refer to selected nodes
3. `buildable` - restrict to tests that ONLY refer to selected nodes (or their ancestors)
4. `empty` - restrict to tests that are only for the selected node and ignore all tests from the attached nodes

i.e., with the default "eager" mode for indirect selection, doing `dbt test` for `model_x` just runs all the tests that are dependent on `model_x`.

You sound like you may want to use one of the other modes (like the `buildable` mode). This can be done on the command line like:
```shell
dbt test -s helper_model --indirect-selection buildable
```

Or you could set this environment-wide with an [environment variable](https://github.com/dbt-labs/dbt-core/issues/7518#issuecomment-1557501256):
```shell
export DBT_INDIRECT_SELECTION=buildable
dbt test -s helper_model
``` 

Wanna give one of those options a shot and let us know how it goes?

**Comment by jaysobel** [comment]
I got learned! Thanks for the help Doug!

**Comment by aymlarr-bamboo** [comment]
Very interesting. I also struggled for a while with a similar issue before I found this report.

I have a generic test that is used only once to test a specific column of a model, which depends on another model. It is declared in the specific model section of the yaml and nowhere else. In that case it seems fair that it would be only called when the first model is built. What's your view on that OOI ?

Anyway, thanks for the fix !
